From 6668b939b81d2f8479afffcd5e0b6dec15bb8e5f Mon Sep 17 00:00:00 2001
From: Yanghan Ye <yyh94306@gmail.com>
Date: Fri, 7 Aug 2026 11:23:21 +0800
Subject: [PATCH] net: airoha: npu: retain the recovery status

Retain the last recovery result and expose airoha_npu_wait_recovery()
so consumers that encounter -EAGAIN while a recovery is running can wait
for it to finish instead of failing again.

Signed-off-by: Yanghan Ye <yyh94306@gmail.com>
---
 drivers/net/ethernet/airoha/airoha_npu.c      |  13 +++++++++++++
 include/linux/soc/airoha/airoha_offload.h     |   7 +++++++
 2 files changed, 20 insertions(+)
--- a/drivers/net/ethernet/airoha/airoha_npu.c
+++ b/drivers/net/ethernet/airoha/airoha_npu.c
@@ -358,6 +358,7 @@
 
 out:
 	spin_lock_bh(&npu->cores[0].lock);
+	npu->recovery_err = err;
 	npu->recovering = false;
 	spin_unlock_bh(&npu->cores[0].lock);
 	mutex_unlock(&npu->recovery_lock);
@@ -371,6 +372,18 @@
 }
 EXPORT_SYMBOL_GPL(airoha_npu_recover);
 
+int airoha_npu_wait_recovery(struct airoha_npu *npu)
+{
+	int err;
+
+	mutex_lock(&npu->recovery_lock);
+	err = npu->recovery_err;
+	mutex_unlock(&npu->recovery_lock);
+
+	return err;
+}
+EXPORT_SYMBOL_GPL(airoha_npu_wait_recovery);
+
 static void airoha_npu_wdt_work(struct work_struct *work)
 {
 	struct airoha_npu_core *core;
--- a/include/linux/soc/airoha/airoha_offload.h
+++ b/include/linux/soc/airoha/airoha_offload.h
@@ -171,6 +171,7 @@
 	/* serialize NPU core recovery */
 	struct mutex recovery_lock;
 	bool recovering;
+	int recovery_err;
 
 	struct airoha_npu_core {
 		struct airoha_npu *npu;
@@ -219,6 +220,7 @@
 struct airoha_npu *airoha_npu_get(struct device *dev);
 void airoha_npu_put(struct airoha_npu *npu);
 int airoha_npu_recover(struct airoha_npu *npu);
+int airoha_npu_wait_recovery(struct airoha_npu *npu);
 
 static inline int airoha_npu_wlan_init_reserved_memory(struct airoha_npu *npu)
 {
@@ -280,6 +282,11 @@
 {
 	return -EOPNOTSUPP;
 }
+
+static inline int airoha_npu_wait_recovery(struct airoha_npu *npu)
+{
+	return -EOPNOTSUPP;
+}
 
 static inline int airoha_npu_wlan_init_reserved_memory(struct airoha_npu *npu)
 {
